### DISCUSSION TITLE
Multiple themes/containers in Theme solution

### DISCUSSION THREAD

**Discussion by ionsource** [oqtane/oqtane.framework#1661] [General] [answered]
Within a single Theme solution, is it possible to have multiple "themes" layouts and "containers"?  There are several things in Oqtane that hint at this possibility, but I can't seem to make this work correctly or find any documentation or examples. Possibly different themes would be defined in the ThemeInfo.cs file?

**Comment by leigh-pointer** [comment]
Have a look at https://github.com/oqtane/oqtane.theme.bootswatch this project should get you going in the right direction.
________________________________
From: ionsource ***@***.***>
Sent: Thursday, September 16, 2021 1:52:15 AM
To: oqtane/oqtane.framework ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [oqtane/oqtane.framework] Multiple themes/containers in Theme solution (Discussion #1661)


Within a single Theme solution, is it possible to have multiple "themes" layouts and "containers"? There are several things in Oqtane that hint at this possibility, but I can't seem to make this work correctly or find any documentation or examples. Possibly different themes would be defined in the ThemeInfo.cs file?

**Comment by sbwalker** [comment, marked as answer]
As @leigh-pointer mentioned the Bootswatch theme is a comprehensive example of what is possible with themes. It takes advantage of the ability to define theme and container settings components which allows a user to configure the behavior. A simpler explanation of the theme system is as follows:

/themefolder - top level folder for a theme - all assets under this folder must use the same namespace
  /themes - all theme components in this folder must inherit from ThemeBase 
    theme1.razor  
    theme2.razor
    theme3.razor
  /containers - all container components in this folder must inherit from ContainerBase
    container1.razor
    container2.razor
  themeinfo.cs - implements ITheme interface and sets properties for Theme

During startup the system will automatically locate all of the theme and container types - no manual registration is required.

(note that Layouts have been deprecated)

**Comment by ionsource** [comment]
Thanks, got this working! Just having one small issue... The containers are repeated for each theme in the "Default Container" and the "Default Admin Containers" dropdowns in the "Appearance" selection. 

If I have two themes and two containers:
- container1.razor
- container2.razor

then "Default Container" and the "Default Admin Containers" dropdowns show:
- container1.razor
- container2.razor
- container1.razor
- container2.razor

Is there a config setting that I missed?

**Reply by sbwalker** [reply]
each theme needs to have its own unique top level folder and namespace

```
/theme1
  /themes
    theme.razor
  /containers
    container.razor
  themeinfo.cs
/theme2
  /themes
    theme.razor
  /containers
    container.razor
  themeinfo.cs
```
